dashboard/lib/sqs/queue_processor.rb

```diff
@@ -0,0 +1,135 @@
+    # stop is called. This does not block and returns immediately after starting the worker
+    # thread, so be sure to make to keep the process running.
+    def start
+      assert_state :initialized, "Can't run in state #{@state}, must be :initialized"
+      @state = :running
+
+      # Kill the process if a thread dies with an unhandled exception. We attempt to catch all possible
+      # exceptions, so this shouldn't happen.
+      Thread.abort_on_exception = true
+      @worker_threads = []
+
+      (1..@config.num_workers_per_processor).each do |i|
+        worker_thread = Thread.new do
+          Thread.current[:name] = "SQS worker #{i}"
+          rate_limiter = RateLimiter.new(@config)
+          poller = Aws::SQS::QueuePoller.new(@config.queue_url)
+
+          # Break out of the polling loop when we leave the running state.
+          poller.before_request do |stats|
+            throw :stop_polling if @state != :running
+          end
+
+          # Wrap the polling loop in an outer exception handler and while loop,
+          # to prevent an unanticipated exception from terminating the thread.
+          while @state == :running
+            begin
+              # Long-poll for messages and handle them until we're told to stop.
+              poller.poll(max_number_of_messages: 10, wait_time_seconds: 20, visibility_timeout: 5) do |sqs_messages|
+                batch_failed = false
+                messages = sqs_messages.map {|sqs_message|
+                  SQS::Message.new(sqs_message.body)
+                }
+                batch_size = sqs_messages.size
+
+                start_time_sec = Time.now.to_f
+                begin
+                  @handler.handle(messages)
+                  @metrics.successes.increment(batch_size)
+                rescue Exception => exception
+                  @logger.warn "Failed on batch of size #{batch_size}"
+                  @metrics.failures.increment(batch_size)
+                  batch_failed = true
+                  @logger.warn exception.message
+                end
+                # Sleep for a bit to make sure we stay below the configured maximum rate. Note that we
+                # pause even if the handler failed because we don't want to exceed the configured rate
+                # even when failures are occuring.
+                delay = rate_limiter.inter_batch_delay(
+                    batch_size: batch_size, elapsed_time_sec: Time.now.to_f - start_time_sec)
+                logger.debug "Sleeping for #{delay} for batch of size #{batch_size}"
+                sleep(delay) if delay > 0
+
+                # Tell SQS to resend the batch if it failed.
+                throw :skip_delete if batch_failed
+              end
+            rescue => exception
+              @logger.warn exception
+            end
+          end # while
+        end
+        @worker_threads << worker_thread
+      end
+    end
+
+    def thread_proc
```

Comment:
This isn't actually used, I will clean it up once the CircleCI tests (hopefully) pass.